### PR TITLE
Implement basic IBC connection handshake in relayer-next

### PR DESCRIPTION
**Pull request by soareschen**
Part of #2925.

## Description

This PR implements the basic IBC connection handshake components in relayer-next. 

The current implementation supports only an end-to-end IBC connection creation, and does not resume connection handshake based on IBC events. It also does not perform additional checks that the created connection on-chain is the same as expected.

Additional features can be defined in future PRs as extra components.

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/relayer-components/src/relay/impls/connection/open_confirm.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+use crate::chain::traits::message_sender::CanSendMessages;
+use crate::chain::traits::queries::status::CanQueryChainHeight;
+use crate::chain::traits::types::height::CanIncrementHeight;
+use crate::relay::traits::chains::HasRelayChains;
+use crate::relay::traits::connection::open_confirm::ConnectionOpenConfirmRelayer;
+use crate::relay::traits::messages::update_client::CanBuildUpdateClientMessage;
+use crate::relay::traits::target::DestinationTarget;
+use crate::std_prelude::*;
+
+pub struct RelayConnectionOpenConfirm;
```
Comment (seanchen1991) [inline crates/relayer-components/src/relay/impls/connection/open_confirm.rs:15]:
```suggestion
/// Component responsible for sending a confirmation message from a source chain
/// to a destination chain in response to an `OPEN_ACK` message confirming that 
/// the source chain received the destination chain's acknowledgement, thereby 
/// completing the IBC connection handshake protocol. 
/// This step is known as the `OPEN_CONFIRM` step.
pub struct RelayConnectionOpenConfirm;
```

File: crates/relayer-components/src/relay/impls/connection/open_handshake.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+use crate::chain::traits::types::ibc::HasIbcChainTypes;
+use crate::relay::traits::chains::HasRelayChains;
+use crate::relay::traits::connection::open_ack::CanRelayConnectionOpenAck;
+use crate::relay::traits::connection::open_confirm::CanRelayConnectionOpenConfirm;
+use crate::relay::traits::connection::open_handshake::ConnectionOpenHandshakeRelayer;
+use crate::relay::traits::connection::open_try::CanRelayConnectionOpenTry;
+use crate::std_prelude::*;
+
+pub struct RelayConnectionOpenHandshake;
```
Comment (seanchen1991) [inline crates/relayer-components/src/relay/impls/connection/open_handshake.rs:11]:
```suggestion
/// Component responsible for performing the IBC connection handshake protocol
/// between a source and a destination chain. Specifically, the `OPEN_TRY`,
/// `OPEN_ACK`, and `OPEN_CONFIRM` steps of the handshake protocol are performed
/// between both chains. Upon successful completion of the handshake protocol,
/// a connection will have been established between both chains.
pub struct RelayConnectionOpenHandshake;
```

File: crates/relayer-components/src/relay/impls/connection/mod.rs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+pub mod bootstrap;
```
Comment (seanchen1991) [inline crates/relayer-components/src/relay/impls/connection/mod.rs:1]:
```suggestion
//! This module contains the traits and types necessary in order to perform the
//! [IBC connection handshake protocol](https://github.com/cosmos/ibc/tree/main/spec/core/ics-003-connection-semantics) 
//! between two IBC-enabled blockchains. The protocol establishes a connection 
//! between the two participating blockchains. This connection between the two 
//! chains is necessary in order for channels to be established between them.
//!
//! Each of the different `open_*` modules, except for `open_handshake`, defines 
//! a component responsible for a single step in the handshake protocol. The 
//! `open_handshake` module wires together the steps that actually constitute
//! the handshake protocol proper. The `open_init` step is not strictly part
//! of the handshake; it merely signals the start of the handshake process. 
//! The `bootstrap` module is the highest level module that wires together the
//! initialization and handshake steps.

pub mod bootstrap;
```

**Review by seanchen1991 — COMMENTED** [review]
I took a stab at documenting each of the different steps of the handshake protocol. Would appreciate a pass-over in case any of it isn't accurate.

**Comment by soareschen** [comment]
> I took a stab at documenting each of the different steps of the handshake protocol. Would appreciate a pass-over in case any of it isn't accurate.

Thanks for the work! I have modified your comments a little and add them directly in a commit.

File: crates/relayer-components/src/relay/impls/connection/mod.rs (hunk 3 above)
Comment (soareschen) [reply crates/relayer-components/src/relay/impls/connection/mod.rs:1]: Technically, the bootstrap function is only used when it is initiated by the relayer user or inside tests. If the connection is initiated by external users, then only the `open_handshake` component is used.

**Pull request merged**